clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -3506,9 +3506,29 @@ class SYCLKernelNameTypeVisitor
             IsInvalid = true;
             return;
           }
-          // Check if the declaration is completely defined within a
-          // function or class/struct.
-          if (Tag->isCompleteDefinition()) {
+
+          // Diagnose used types without complete definition only on host
```

Comment:
Can we update the comment to something like - 

```
Diagnose used types without complete definition i.e. 
 int main() {
   class KernelName1;
   parallel_for<class KernelName1>(..);
 }

This case can only be diagnosed during host compilation because the integration header is required to
distinguish between the invalid code (above) and the following valid code:

int main() {
  parallel_for<class KernelName2>(..);
 }

The device compiler 'forward declares' both KernelName1 and KernelName2 in the integration header as
::KernelName1 and ::KernelName2. The problem with the former case is the additional declaration class
KernelName1 in non-global scope. Lookup in this case will resolve to ::main::KernelName1 (instead of
::KernelName1). Since this is not visible to runtime code that submits kernels, this is invalid.
```

WDYT? I think this is a bit more clear. Please correct me if I've misunderstood something here.

Reply:
Yeah, that is definitely better than current version. Will do, thank you very much!

Reply:
No problem! Maybe class KernelName1 should also be in quotes/comma or something so sentence reads better. Like - 

`additional declaration, class KernelName1, in non-global scope`

Reply:
Yep. Done. Though I omitted the quotes around 'forward declares' because integration header really does that :) Please let me know if that is ok.

Reply:
LGTM! Thanks!